staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -432,3 +446,42 @@ func (r *Reflector) setLastSyncResourceVersion(v string) {
 	defer r.lastSyncResourceVersionMutex.Unlock()
 	r.lastSyncResourceVersion = v
 }
+
+// relistResourceVersion determines the resource version the reflector should list or relist from.
+// Returns either the lastSyncResourceVersion so that this reflector will relist with a resource
+// versions no older than has already been observed in relist results or watch events, or, if the last relist resulted
+// in an HTTP 410 (Gone) status code, returns "" so that the relist will use the latest resource version available in
+// etcd via a quorum read.
+func (r *Reflector) relistResourceVersion() string {
+	r.lastSyncResourceVersionMutex.RLock()
+	defer r.lastSyncResourceVersionMutex.RUnlock()
+
+	if r.isLastSyncResourceVersionGone {
+		// Since this reflector makes paginated list requests, and all paginated list requests skip the watch cache
+		// if the lastSyncResourceVersion is expired, we set ResourceVersion="" and list again to re-establish reflector
+		// to the latest available ResourceVersion, using a consistent read from etcd.
+		return ""
```

Comment:
I hope that with "watch bookmarks" it won't hit us seriously, but it has some potential to break scalability (that said, I agree that we need  it for correctness). I'm just pointing to something that we should look at and if it will appear to happen more than rarely, consider some tuning of watch bookmarks features (or figure out if other things can be done).

Reply:
is the long-term goal to use watch heartbeat on etcd 3.4 to ensure lists at `resourceVersion=0` that are served from the watch cache only return once their data is current with etcd as of the time of the request? would we change this implementation at that point?

Reply:
I was thinking we'd handle it by making it possible to serve RV="" as a consistent read from watch cache (instead of always hitting etcd for the entire list request), and not guarantee consistency for RV="0".

This should be possible to do. The only needed from etcd to serve consistent reads is which RV is consistent, and then the read can be served from that RV from cache, if the cache has it. etcd has a progress notify operation (progress events are basically bookmarks, but for etcd watch). I've got a sketch of how this might all work, but need to discuss it more broadly before we decide if it's something we'd want to implement. I have a doc I can dust off and circulate.